### PR TITLE
fix(Core/Loot): Only Paladins, Warriors and Shamans should be able to…

### PR DESCRIPTION
**Pull request by r0m1ntik**
… roll Need on shields

## Changes Proposed:
-  Loot shields only for class (paladin, warrior, shaman)

## Issues Addressed:
- Closes [4858](https://github.com/azerothcore/azerothcore-wotlk/issues/4858)
- Closes [chromiecraft/chromiecraft#190](https://github.com/chromiecraft/chromiecraft/issues/190) 

## Tests Performed:
Paladin:
Снимок экрана 2021-03-16 в 12 38 19
Lock:
Снимок экрана 2021-03-16 в 12 38 32
Снимок экрана 2021-03-16 в 12 38 35

## How to Test the Changes:

1. Enter group
2. Party loot or need before greed
3. Shield drops
4. Roll need on a shield as a other classes.

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Mitradis** [comment]
*joke about Hunters

**Comment by Hacki95** [comment]
Is the Lock not seing the shield at all? Shouldn't he atleast be able to roll greed?

**Comment by AbraKabastard** [comment]
Yes, Please look at putting your code in 
```InventoryResult Player::CanRollForItemInLFG``` in Player.cpp

File: src/server/game/Loot/LootMgr.cpp [hunk 1]
```diff
@@ -412,6 +412,10 @@ bool LootItem::AllowedForPlayer(Player const* player) const
     if (!(pProto->FlagsCu & ITEM_FLAGS_CU_IGNORE_QUEST_STATUS) && ((needs_quest || (pProto->StartQuest && player->GetQuestStatus(pProto->StartQuest) != QUEST_STATUS_NONE)) && !player->HasQuestForItem(itemid)))
         return false;
 
+    // check for shields
+    if ((pProto->SubClass == ITEM_SUBCLASS_ARMOR_SHIELD) && !(player->getClass() == CLASS_PALADIN || player->getClass() == CLASS_WARRIOR || player->getClass() == CLASS_SHAMAN))
+        return false;
```
Comment (FrancescoBorzi) [inline src/server/game/Loot/LootMgr.cpp:417]: @r0m1ntik could you please add brackets?

**Comment by r0m1ntik** [comment]
@Hacki95  @AbraKabastard  can you check it ?

**Comment by Stifler82** [comment]
**Related to:** https://github.com/chromiecraft/chromiecraft/issues/190

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by Hacki95** [comment]
@FrancescoBorzi could you give me one more day to check if other classes are able to roll need on a shield? I am pretty sure it was like this I just need to find a video.

**Comment by Si1ker** [comment]
> @FrancescoBorzi could you give me one more day to check if other classes are able to roll greed on a shield? I am pretty sure it was like this I just need to find a video.

Let me know when finished. (tag)

**Comment by Hacki95** [comment]
> 
> 
> > @FrancescoBorzi could you give me one more day to check if other classes are able to roll greed on a shield? I am pretty sure it was like this I just need to find a video.
> 
> Let me know when finished. (tag)

Yes I found something which proves that this PR is not working right.

https://wowpedia.fandom.com/wiki/Need_Before_Greed

grafik

How its supposed to work: Lets say we got a group with: 1 Warrior Tank, 1 Priest Heal, 2 DPS Rogues, 1 Mage

1 Shield drops: Only Warrior can roll need, for the other members need is greyed out and they can only click greed.
If the warrior selects need, he gets it as he is the only one rolling need. If he rolls greed, all the other people who rolled greed (you can also pass) get to roll for the item.

This works for every other type of armor/weapon too. Lets say a Dagger drops: Everyone in the group can wear daggers so everyone can roll need.

Leather armor drops: Leather is for the rogues, so only those two can roll need. All the other people can roll greed.

And so on, it works like this for every single item type not only shields.

If it would work like your PR then classes like mages would have a huge disadvantage for farming runs because for them only cloth and a few weapons like daggers and staffs would drop for them.

**Comment by Si1ker** [comment]
Yeah I think the solution is wrong-addressed (thanks for the effort) , the issue is "all classes can roll need for shield" , ok the solution would be "only paladins & warriors can roll need", all the other only greed.

**Comment by Hacki95** [comment]
> 
> 
> Yeah I think the solution is wrong-addressed (thanks for the effort) , the issue is "all classes can roll need for shield" , ok the solution would be "only paladins & warriors can roll need", all the other only greed.
> cc @r0m1ntik

Don't forget Shamans!

**Comment by AbraKabastard** [comment]
See Player.cpp 
```void Group::NeedBeforeGreed(Loot* loot, WorldObject* lootedObject)```
will eventually hit this line 
```SendLootStartRollToPlayer(60000, lootedObject->GetMapId(), p, p->CanRollForItemInLFG(item, lootedObject) == EQUIP_ERR_OK, *r);``` That fourth parameter in SendLootStartRollToPlayer is a boolean canNeed. Your code should hypothetically fix the problem IF the group is an LFG group and has loot set to need before greed. If you're in a regular party the first line of CanRollForItemInLfg is ```if (!GetGroup() || !GetGroup()->isLFGGroup())``` which will return EQUIP_ERR_OK in a non LFG formed group.

I can't find ANY function that handles the ability to need / greed based on this kind of thing but reading https://wowpedia.fandom.com/wiki/Need_Before_Greed , specifically :
```
With patch 3.3.0, the Need Before Greed system was redesigned to recognize gear appropriate for a class:

    The class must be able to equip the item.
    The class must have a spec that uses the item's primary stat (strength, agility or intellect if one of these 
    is present; stamina otherwise).
    Classes are limited to their dominant armor type (ex. paladins for plate).

For items other than equipment, other restrictions apply; you can't roll Need on:

    Mounts or companions you already have.
    Recipes you don't have the required professions for, including tradable recipes for items that you need.
     [Chaos Orb] if you don't have at least 500 in a profession that uses it.
```

Do we need a can roll need function that doesn't take into account LFG for NeedBeforeGreed? I would think so but I'm spitballing and after some of your thoughts.  I think we could repurpose CanRollForItemInLFG because its only used in NeedBeforeGreed code and take out the LFG checks, or atleast bracket them for this little block : 
```c++
    // check if looted object is inside the lfg dungeon
    Map const* map = lootedObject->GetMap();
    if (!sLFGMgr->inLfgDungeonMap(GetGroup()->GetGUID(), map->GetId(), map->GetDifficulty()))
        return EQUIP_ERR_OK;
```
sorry for wall of text

**Comment by Hacki95** [comment]
I also remembered one more thing, the rule that only certain classes can need on shields is only applicable in the RDF! In the open world with a normal group it should be possible for anyone to roll need on a shield.

**Comment by AbraKabastard** [comment]
> I also remembered one more thing, the rule that only certain classes can need on shields is only applicable in the RDF! In the open world with a normal group it should be possible for anyone to roll need on a shield.

That wiki page says otherwise for need before greed

**Comment by Stifler82** [comment]
@r0m1ntik 
Can you add this to "Closes" pls?
https://github.com/chromiecraft/chromiecraft/issues/190

**Comment by FrancescoBorzi** [comment]
what's the status of this? should it be merged or needs something else?

**Comment by Hacki95** [comment]
> 
> 
> what's the status of this? should it be merged or needs something else?

Should not be merged. It needs to work differently, shields should still drop for the other members but they should only be able to greed it.

**Comment by FrancescoBorzi** [comment]
should this be closed then? or what?  @r0m1ntik

**Comment by UltraNix** [comment]
That code looks valid to me and I agree with it, but I didn't test it.

**Review by Helias — APPROVED** [review]

**Comment by Helias** [comment]
@azerothcore/testers

**Comment by Sedatyf** [comment]
I don't know what's the status of this, but I tried on my local core and it didn't worked. I was in wailing cavern, killed Verdan the Everliving who drops Seedcloud Buckler, and I could roll need as a hunter

My group was in Need Before Greed and I also tried in Group Loot and I had the same behavior

image

**Comment by Hacki95** [comment]
I also don't get why this was approved, I said several times this mechanic is not Blizzlike this way.

**Comment by Sedatyf** [comment]
I didn't read through all comments to be honest. So we drop this PR?

**Comment by FrancescoBorzi** [comment]
> I also don't get why this was approved, I said several times this mechanic is not Blizzlike this way.

it is approved only for code review as the code looks fine (as in codestyle), it's not [To Be Merged]

**Comment by Sedatyf** [comment]
So what is your POV on this @FrancescoBorzi 
Do we keep this or not?

**Comment by claudiodfc** [comment]
Hacki is right, this PR not only doesn't fix the issue but creates another. Until then leave open.

**Comment by UltraNix** [comment]
As I'm quite sure it's a good code and should work, please retest it.

**Comment by syssneck** [comment]
I can't test it, I'm getting merge conflicts. Can someone merge master into this PR again, please?

**Comment by UltraNix** [comment]
I will make a new PR as @r0m1ntik seems to be gone.

**Pull request closed without merging**